[APM] Fix indefinite loading state in agent settings for unauthorized user roles

ogupte opened the pull request:
Addresses #43326 (in conjunction with https://github.com/elastic/elasticsearch/pull/46423)

- displays message to user when a failure occurred in settings initialization
- fixes incorrect settings link path
- moves agent config index creation to plugin setup step

<img width="851" alt="Screen Shot 2019-09-08 at 10 52 29 PM" src="https://user-images.githubusercontent.com/1967266/64506466-8deec280-d28c-11e9-9d52-d18ac809c0a1.png">

ogupte committed:
- handle unauthorized error to return empty list of agent settings
- display to user when a failure occurred in settings initialization
- fix incorrect settings link path

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1501/)

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/Links/apm/SettingsLink.tsx:
```diff
@@ -7,7 +7,7 @@ import React from 'react';
 
 const SettingsLink = (props: APMLinkExtendProps) => {
-  return <APMLink path="/" {...props} />;
+  return <APMLink path="/settings" {...props} />;
```
woops 😅 thanks for catching this.

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/app/Settings/SettingsList.tsx:
```diff
@@ -294,10 +321,9 @@ export function SettingsList() {
-        {status === 'success' && !hasConfigurations ? (
-          emptyStatePrompt
-        ) : (
+        {status === 'failure' ? (
```
Not a super big fan of nested ternaries in terms of readability, might be just me though.

sqren replied:
I agree. I just didn't have any better suggestions :)

dgieselaar commented on x-pack/legacy/plugins/apm/server/routes/settings.ts:
```diff
@@ -20,13 +20,19 @@ import { transactionSampleRateRt } from '../../common/runtime_types/transaction_
   path: '/api/apm/settings/agent-configuration',
-  handler: async req => {
-    await createApmAgentConfigurationIndex(core.http.server);
+  handler: async (req, _, h) => {
```
are we using `h`? can't see it being used below.

ogupte replied:
my mistake for leaving this in

dgieselaar commented on x-pack/legacy/plugins/apm/server/routes/settings.ts:
```diff
@@ -20,13 +20,19 @@ import { transactionSampleRateRt } from '../../common/runtime_types/transaction_
+      const setup = await setupRequest(req);
+      return await listConfigurations({ setup });
+    } catch (error) {
+      if (error.statusCode === 403) {
```
If we catch 403 authorization errors, how does the user ever see `failurePrompt`? What other status codes should we expect here? just 500?

ogupte replied:
failurePrompt is now make more generic and the statusCode check was removed

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/SettingsList.tsx:
```diff
@@ -173,6 +173,33 @@ export function SettingsList() {
+              'xpack.apm.settings.agentConf.configTable.failurePromptText',
+              {
+                defaultMessage:
+                  'Whoops! The settings management failed to initialize. This might be because the user that is logged in does not have permission to read or write the APM agent configuration index. The next time a user with adequate privileges loads this screen, Kibana will complete this step.'
```
This text is still confusing. Our implementation details wrt creating the index lazily should not shine through to the end-user. 

If we assume the apm_user has read/write access to the agent configuration index we have a couple of use cases:
 - An unknown backend error occurred (network glitch, database corruption, anything). We do not handle this in the rest of APM UI except for showing the automatic toast that the request failed. Feel free to explicitly handle the failure state. In that text the text needs to be something vague like "The list of agent configurations could not be fetched".
 - The user has neither read or write permission: they'd get the above error. That's better than most other places in APM UI.
 - The user has read permission but not write permission: the user should be able to see the list (and shouldn't see errors regardless if the index could not be created)
 - The user has read and write permissions: the user should be able to see the list

sqren commented on x-pack/legacy/plugins/apm/server/routes/settings.ts:
```diff
@@ -20,13 +20,19 @@ import { transactionSampleRateRt } from '../../common/runtime_types/transaction_
+      return await listConfigurations({ setup });
+    } catch (error) {
+      if (error.statusCode === 403) {
+        // return an empty list if user is not authorized
```
I think you should only do this for `createApmAgentConfigurationIndex`. I don't think we should swallow errors for `listConfigurations` this way.

sqren replied:
This assumes that the apm_user as a minimum has read access to the index.

ogupte replied:
The latest commit removed all the assumptions and displays a message to the user that a failure occurred when loading settings.

Index creation uses `callWithInternalUser` during plugin setup, so any index creation failures should not occur when a user loads the settings screen. but it will error if the user does not have permission to read from the .apm-agent-configuration index. as per your list of use cases above.

When testing with a user with the role `apm_user`, i discovered that it does not have read permissions on this index. Since it's not appropriate to call `callWithInternalUser` for querying this index, it will show the failure message. The alternative is to give the `apm_user` role the permission to read the agent config index. I created a PR to update the role, however it looks like it will not get approval (https://github.com/elastic/elasticsearch/pull/46423).

ogupte committed:
- moved agent config index creation to the plugin setup step
- make failure text reflect a general failure to load settings

ogupte commented on x-pack/legacy/plugins/apm/public/components/app/Settings/SettingsList.tsx:
```diff
@@ -294,19 +313,21 @@ export function SettingsList() {
+          ) : (
+            emptyStatePrompt
+          )
+        ) : null}
```
The previous `status === 'success' && !hasConfigurations` logic was rending `emptyStatePrompt` when `status` was `loading` so the new nested logic makes sure that nothing is rendered at that time. This is especially jarring if there's a failure message since it used to flash the `emptyStatePrompt` immediately prior to the failure message.

I'm open to alternatives for the nested ternary. :)

sqren replied:
What about a new component:
```ts
function Foo(status) {
  if (status === 'failure') {
    return failurePrompt;
  }

  if (status === 'success') {
    if (!hasConfigurations) {
      return emptyStatePrompt;
    }

    return (
      <ManagedTable
        noItemsMessage={<LoadingStatePrompt />}
        columns={COLUMNS}
        items={data}
        initialSortField="service.name"
        initialSortDirection="asc"
        initialPageSize={50}
      />
    );
  }
}
```

ogupte replied:
created a new component for this called `SettingsList` (the old SettingsList is now just Settings/index since it renders the settings page.

ogupte commented on x-pack/legacy/plugins/apm/server/new-platform/plugin.ts:
```diff
@@ -7,12 +7,13 @@
   public setup(core: InternalCoreSetup) {
     createApmApi().init(core);
-
+    createApmAgentConfigurationIndex(core);
```
This issue made it clear to me that index creation belongs in this setup step.

sqren replied:
I seem to recall we did that initially and then we were asked not to do it during startup. Or was that only for the index pattern?

sqren replied:
If this is okay for platform I'm +1. Seems like the right place to do it 👍

ogupte replied:
it was the index pattern before because it prevented the welcome screen, but this index doesn't have the same effect.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1822/)

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/SettingsList.tsx:
```diff
@@ -173,6 +173,25 @@ export function SettingsList() {
+              'xpack.apm.settings.agentConf.configTable.failurePromptText',
+              {
+                defaultMessage:
+                  'The list of agent configurations could not be fetched'
```
I know I'm going back and forth on this but my assumptions have changed a little.
If we do not expect the `apm_user` to have read access, and we therefore expect many users to see this issue because of permission problems, it might make sense to hint about that.

Eg. "The list of agent configurations could not be fetched. Your user may not have the sufficient permissions."

To be clear, I was more against the fact that the user was seeing an error for something they shouldn't be concerned about (they were seeing an issue if they didn't have write access, even if they just wanted to view/read the list).

sorenlouv reviewed (approved):
Looks good. Mostly nits.

ogupte committed:
- Rename Settings/SettingsList -> Settings/index
- Add Settings/SettingsList to render only the list/messaging
- Mention permissions issue in failure message

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1987/)

sqren commented:
This works as expected. However a new issue has been opened related to this: https://github.com/elastic/kibana/issues/45610
